### DISCUSSION TITLE
large scale heterogeneous graph classification

### DISCUSSION THREAD

**Discussion by leethu2012** [pyg-team/pytorch_geometric#5824] [Q&A]
I am building a classifier for heterogeneous graphs. The graphs are too big  and cause GPU OOM problem during training.  
Can you give some advice for this problem? 
I wonder whether sampling methods, such as HGTLoader or NeighborLoader, which are used for node classification, can be used for large scale heterogeneous graph classification?
Thanks for any advice!

**Comment by rusty1s** [comment]
Do your graph don‘t even fit with a batch size of 1? One way you could solve it is to run `data.subgraph` on a random node subset before moving data to the GPU. This way, you can reduce the graph size.

**Reply by leethu2012** [reply]
Yes, GPU got OOM even setting batch size to 1 during training.
Thank for your advice, I will try data.subgraph out. What I still concern about are:

1. To run data.subgraph on a random node subset , maybe I can keep the whole graph data in cpu memory, then use something like RandomNodeSampler to load  node subset to gpu and training iteratively, just as [this example](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/ogbn_proteins_deepgcn.py).
2. Since [this example](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/ogbn_proteins_deepgcn.py) is for node classfcation, my question is, sampling methods, such as HGTLoader,  RandomNodeSampler or NeighborLoader can be used for graph classification? Or when used, will the accuracy of the training be affected?
3. is [pyg_autoscale ](https://github.com/rusty1s/pyg_autoscale) another option? What i should do if I want to use autoscale for heterogeneous graph classification? Can these three ways in the [docs](https://pytorch-geometric.readthedocs.io/en/latest/notes/heterogeneous.html) be applied with autoscale to create models on heterogeneous graph data?
![image](https://user-images.githubusercontent.com/29704017/197909668-c4ac4f24-e254-4478-a0fe-2d8fa6a70863.png)

**Reply by rusty1s** [reply]
1. Yes, you would still need to hold the big graph in memory and just reduce it in order to be able to put it onto the GPU. `RandomNodeSampler` will perform indeed something similar here.
2. Yeah, this methods are mostly for node classification. I do not recommend using them for graph classification.
3. `pyg_autoscale` is also more targeted towards node classification. It can potentially be used for graph classification as well, but this is not yet integrated into the code base. It also does not support heterogeneous graphs at the moment :(

**Reply by jiaqian** [reply]
Hi Matthias,
Does `RandomNodeSampler` only handle `Data` object instead of `HeteroData`?

**Reply by rusty1s** [reply]
Yes, currently `RandomNodeSampler` can only handle `Data` objects.

**Reply by jiaqian** [reply]
is there any plan in the new release to cover the heterogeneous graph or any other way that I can generate subgraphs on **heterogeneous** graph? Thanks.

**Reply by rusty1s** [reply]
You should be able to do so via `hetero_data.subgraph(...)`.

**Reply by jiaqian** [reply]
with `hetero_data.subgraph()`, I need to supply the subset_dict ? In my case, I have a very big heterogeneous graph and I want to split it into subgraphs and do the inference on each subgraph (similar to `ClusterData`, but I know it does not handle heterogeneous graph).

**Reply by rusty1s** [reply]
Yes, you would be responsible for creating the `subset_dict` on your own, e.g., via
```python
subset_dict = {}
for node_type in data.node_types:
    perm = torch.randperm(data[node_type].num_nodes
    subset = perm[:10000]  # take 10,000 random nodes per node type
    subset_dict[node_type] = subset

**Reply by rusty1s** [reply]
Will also work on `HeteroData` support for the random node loader.

**Reply by jiaqian** [reply]
ok, I actually could not provide `subset_dict` since I don't want to randomly split the graph. But I think I can make the graph homogeneous in the first place and use `SparseTensor.partition()` to create a list of subgraphs ([A_1,A_2,...,A_n]) and take each of them as a heterogeneous graph later to run models. Thank you so much.

**Reply by rusty1s** [reply]
Yes, this works as well.

**Reply by rusty1s** [reply]
Integrated into `RandomNodeLoader` via #6007.
